csrc/optimization/exact_mapped_extent_substitution.cpp

```diff
@@ -0,0 +1,90 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <debug.h>
+#include <ir/utils.h>
+#include <optimization/exact_mapped_extent_substitution.h>
+#include <options.h>
+#include <root_domain_map.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+void exactMappedExtentSubstitution(Fusion* fusion) {
+  // map non-const extents to const extents
+  std::unordered_map<Val*, Val*> replacement_map;
+
+  const auto mapped_sets = ExactRootDomainMap(fusion).getMappedSets();
+  // Loop over each exact root domain set
+  for (const auto& set_ptr : mapped_sets.disjointSets()) {
+    // (1) pick a const extent
+    // (2) if no const extent, pick the var with the lowest name()
+    Val* const_extent = nullptr;
+    Val* lowest_val = nullptr;
+    bool can_substitute = true;
+    for (auto id : *set_ptr) {
+      // If one of the domains is not a root domain, we cannot substitute.
+      // e.g. in this set { iS11{( i0 * i2 )}rf; iS12{( i0 * i2 )}; iS14{i3} }
+      // we can't substitute {i0 * i2} with {i3}, otherwise,
+      // ValidateDomainEquivalence fails. If we really want to substitute, we
+      // may need to skip ValidateDomainEquivalence.
```

Comment:
This might just be because of how we are doing the replacement currently using replaceValue. It seems like this condition would be undesirable since we might have just one derived ID, say a padded dimension in an exact group that contains a constant extent and this condition would prevent substitution for the whole group. We could decline to substitute just the extents of those derived IDs, or look into why we are not preserving definitions during Val replacement.

Reply:
Thanks! Let me check if this approach can also pass all the tests.